zjit/src/codegen.rs

```diff
@@ -917,6 +917,41 @@ fn gen_new_array(
     new_array
 }
 
+/// Compile a new hash instruction
+fn gen_new_hash(
+    jit: &mut JITState,
+    asm: &mut Assembler,
+    elements: &Vec<(InsnId, InsnId)>,
+    state: &FrameState,
+) -> Option<lir::Opnd> {
+    // Save PC
+    gen_save_pc(asm, state);
+    gen_save_sp(asm, state.stack().len());
+
+    // Spill the virtual stack and the locals of the caller onto the stack
+    // TODO: Lazily materialize caller frames on side exits or when needed
+    asm_comment!(asm, "spill locals and stack");
+    for (idx, &insn_id) in state.locals().enumerate() {
+        asm.mov(Opnd::mem(64, SP, (-local_idx_to_ep_offset(jit.iseq, idx) - 1) * SIZEOF_VALUE_I32), jit.get_opnd(insn_id)?);
+    }
+    for (idx, &insn_id) in state.stack().enumerate() {
+        asm.mov(Opnd::mem(64, SP, idx as i32 * SIZEOF_VALUE_I32), jit.get_opnd(insn_id)?);
+    }
+
+    asm_comment!(asm, "call rb_hash_new");
+    let cap: ::std::os::raw::c_long = elements.len().try_into().expect("Unable to fit length of elements into c_long");
+    let new_hash = asm_ccall!(asm, rb_hash_new_with_size, lir::Opnd::Imm(cap));
+
+    for (key_id, val_id) in elements.iter() {
+        let key = jit.get_opnd(*key_id)?;
+        let val = jit.get_opnd(*val_id)?;
+        asm_comment!(asm, "call rb_hash_aset");
+        asm_ccall!(asm, rb_hash_aset, new_hash, key, val);
```

Comment:
Having a loop here could make ZJIT slower than the interpreter. Could you use the same function that the interpreter uses, `rb_hash_bulk_insert`? YJIT uses that too.

Reply:
@k0kubun Out of curiosity, how would we handle it if a user-defined key throws an exception while inserting multiple keys at once?

Reply:
When something raises an exception in the middle of a C call, it would longjmp to a frame that did setjmp. At that point, this JIT frame will be expired, so there's nothing to handle in the JIT code. Thanks to `gen_prepare_non_leaf_call`, the exception handler could use the stack spilled by the JIT code to take over the execution on the interpreter when there's a catch table for the ISEQ.